src/vsr/replica.zig

```diff
@@ -7173,6 +7201,27 @@ pub fn ReplicaType(
             }
         }
 
+        fn send_prepare_oks_after_checkpoint(self: *Self) void {
+            assert(self.status == .normal or self.status == .view_change or
+                (self.status == .recovering and self.solo()));
+
+            const op_checkpoint_trigger =
+                vsr.Checkpoint.trigger_for_checkpoint(self.op_checkpoint()).?;
+            assert(self.commit_min == op_checkpoint_trigger);
+
+            var op = op_checkpoint_trigger + constants.pipeline_prepare_queue_max + 1;
+            while (op <= self.op) : (op += 1) {
+                // We may have breaks or stale headers in our uncommitted chain here. However:
+                // * being able to send what we have will allow the pipeline to commit earlier, and
+                // * the primary will drop any prepare_ok for a prepare not in the pipeline.
+                // This is safe only because the primary can verify against the prepare checksum.
+                if (self.journal.header_with_op(op)) |header| {
+                    self.send_prepare_ok(header);
+                    defer self.flush_loopback_queue();
+                }
+            }
+        }
```

Comment:
Extract common code between this function and the one bellow into

```
fn send_prepare_oks_from(self: *Self, op: u64)
```

I am wondering if for `send_prepare_oks_after_checkpoint` we could start from `@max(commit_max, trigger + pipeline + 1)`? That is, if we already know that the stuff is committed, no need to re-ack it.

Reply:
Great to get your green light on this, I was _about_ to ask you about this deduplication but decided against it because the two scenarios have distinct things we can assert! Let me lift those up to the callsite if they don't exist already.

Reply:
Also, nice observation re sending duplicate prepare_oks. I've incorporated both comments in https://github.com/tigerbeetle/tigerbeetle/pull/2236/commits/d77bf8d17a44d74addde2fdd224981a5c000f36a!